File: Prism.Scripting/src/Prism/Math/Transform.cs
using System;
using System.Runtime.InteropServices;

namespace Prism
{
    [StructLayout(LayoutKind.Sequential)]
    public struct Transform : IEquatable<Transform>
    {
        public Vector3 Position;
        public Vector3 Rotation; // euler angles in degrees
        public Vector3 Scale;
        public Vector3 Up;       // filled by C++
        public Vector3 Right;    // filled by C++
        public Vector3 Forward;  // filled by C++

        public Transform(Vector3 position, Vector3 rotation, Vector3 scale,
                         Vector3 up, Vector3 right, Vector3 forward)
        {
            Position = position;
            Rotation = rotation;
            Scale = scale;
            Up = up;
            Right = right;
            Forward = forward;
        }

        // Compose two transforms: result = a * b  (apply b, then a)
        public static Transform operator *(Transform a, Transform b)
        {
            Quaternion aq = new(a.Rotation * MathF.PI / 180f);
            Quaternion bq = new(b.Rotation * MathF.PI / 180f);
            Quaternion cq = aq * bq;

            Transform result;
            result.Position = a.Position + aq * (a.Scale * b.Position);
            result.Rotation = cq.EulerAngles() * (180f / MathF.PI);
            result.Scale = new Vector3(
                a.Scale.X * b.Scale.X,
                a.Scale.Y * b.Scale.Y,
                a.Scale.Z * b.Scale.Z
            );
            result.Up = aq * Vector3.Up;
            result.Right = aq * Vector3.Right;
            result.Forward = aq * Vector3.Forward;
            return result;
        }

        public override bool Equals(object? obj) => obj is Transform other && Equals(other);
        public bool Equals(Transform right)
            => Position == right.Position && Rotation == right.Rotation && Scale == right.Scale;
        public override int GetHashCode() => HashCode.Combine(Position, Rotation, Scale);
        public static bool operator ==(Transform left, Transform right) => left.Equals(right);
        public static bool operator !=(Transform left, Transform right) => !(left == right);
    }
}

File: PrismEditor/Assets/scripts/Python/Prism/Core/Transform.py
import PrismNative as _Prism
from Prism.Math.Vector3 import Vector3


class Transform:
    """纯数据容器，所有字段由 C++ 引擎填充。"""

    def __init__(self, position: Vector3, rotation: Vector3, scale: Vector3,
                 up: Vector3 = None, right: Vector3 = None, forward: Vector3 = None):
        self.Position = position
        self.Rotation = rotation
        self.Scale = scale
        self.Up = up or Vector3(0.0, 1.0, 0.0)
        self.Right = right or Vector3(1.0, 0.0, 0.0)
        self.Forward = forward or Vector3(0.0, 0.0, -1.0)
